src/coins.c

```diff
@@ -0,0 +1,167 @@
+#include "internal.h"
+
+#include <include/wally_coinselection.h>
+
+typedef struct value_remaining {
+    uint64_t remaining;
+    uint64_t value;
+} value_remaining_t;
+
+/*
+ * Coin selection for assets is much simpler than the policy asset L-BTC.
+ *
+ * Assumptions/Limits:
+ * - We can only have up to 252 + 1 change asset outputs, allowing 1 fee and
+ *   1 L-BTC change output to remain within the Elements 255 output limit.
+ * - We limit the number of asset inputs in the solution to 256 for simplicity.
+ *   Finding such a solution other than a trivial one in computationally
+ *   feasible time is highly improbable anyway.
+ * - We assume the outputs will be blinded.
+ * - The same asset is usually/mostly received to the same wallet/on the same
+ *   address type and so we can ignore effective value in favour of absolute
+ *   value.
+ * - Dust is not meaningful for assets, since we cannot know their value
+ *   relative to L-BTC. We nevertheless attempt to use only the largest UTXOs.
+ * - Privacy is somewhat protected by blinding of the asset and value, so we
+ *   do not care about spending the same addresses together.
```

Comment:
If there are outputs with the same scriptpubkey, not spending them together allows to link transactions, which is a privacy issue.

However in elements if we spend all utxos (of any asset) which have the same scriptpubkey, an attacker might cheaply send assets without value to the target, which will have to waste fees for the extra tx inputs and (perhaps) outputs.

OTOH if the wallet has only assets which have some value, then spending all utxos with the same scriptpubkey is better from a privacy perspective.

We might not want to do it for complexity and costs reasons though.

Reply:
> spend all utxos (of any asset) 

This will never be a good idea AFAICS. Its one thing to scoop up related inputs, but a very different proposition to create new outputs. I think the best cause of action for users receiving unwanted/spam assets is to blacklist those UTXOs or ignore them (since they shouldn't be able to be spent accidentally).

> spending all utxos with the same scriptpubkey is better from a privacy perspective

I think this case is rare enough that it can be handled outside of the the exposed wally call if the caller wishes to, e.g. if the caller determines that some number of utxos share a spk:

- First try a normal selection. If a solution is found that includes either all or none of the same-spk inputs, then use it.
- Otherwise, combine the same-spk inputs together into a single value, and perform a selection with io_ratio increased.

I'm expecting there may be alternate selection algorithms in the future, some of which could take extra data like the spk or a tag to mark inputs that should be spent together. However I don't want to complicate the standard logic for this scenario.